src/paillier.rs

```diff
@@ -272,4 +306,133 @@ mod test {
         }
     }
+
+    fn random_message(
+        rng: &mut (impl CryptoRng + RngCore),
+        encryption_key: &PaillierEncryptionKey,
+    ) -> BigNumber {
+        let range: BigNumber = encryption_key.n() / 2;
+        random_bn_plusminus(rng, &range)
+    }
+
+    #[test]
+    fn paillier_encryption_works() {
+        let mut rng = get_test_rng();
+        let (decryption_key, _, _) = PaillierDecryptionKey::new(&mut rng).unwrap();
+        let encryption_key = decryption_key.encryption_key();
+
+        for _ in 0..100 {
+            let msg = random_message(&mut rng, &encryption_key);
+
+            // Encryption on good inputs doesn't fail
+            let enc_result = encryption_key.encrypt(&mut rng, &msg);
+            assert!(enc_result.is_ok());
+            let (ciphertext, _) = enc_result.unwrap();
+
+            // Decryption on good inputs doesn't fail
+            let dec_result = decryption_key.decrypt(&ciphertext);
+            assert!(dec_result.is_ok());
+            let decrypted_msg = dec_result.unwrap();
+
+            // Decrypted message matches original
+            assert_eq!(msg, decrypted_msg);
+        }
+    }
+
+    #[test]
+    fn pailler_decryption_requires_correct_key() {
+        let mut rng = get_test_rng();
+        let (decryption_key, _, _) = PaillierDecryptionKey::new(&mut rng).unwrap();
+        let encryption_key = decryption_key.encryption_key();
+
+        let msg = random_message(&mut rng, &encryption_key);
+        let (ciphertext, _) = encryption_key.encrypt(&mut rng, &msg).unwrap();
+
+        let (wrong_decryption_key, _, _) = PaillierDecryptionKey::new(&mut rng).unwrap();
+        let decryption_result = wrong_decryption_key.decrypt(&ciphertext);
+        assert!(decryption_result.is_err() || decryption_result.unwrap() != msg);
+
+        assert_eq!(decryption_key.decrypt(&ciphertext).unwrap(), msg);
+    }
+
+    #[test]
+    fn pailler_encryption_requires_input_in_Zn() {
+        // Specifically, the integers mod N must be in the interval around 0.
+        // So, inputs in the range [-N/2, N/2] are acceptable, but not [0, N).
+        let mut rng = get_test_rng();
+        let (decryption_key, _, _) = PaillierDecryptionKey::new(&mut rng).unwrap();
+        let encryption_key = decryption_key.encryption_key();
+
+        // In the acceptable range, 0 is allowed
+        assert!(encryption_key.encrypt(&mut rng, &BigNumber::zero()).is_ok());
+
+        // Test a number in between N/2 and N (both + and -)
+        let too_big = (encryption_key.n() / 3) * 2;
+        assert!(encryption_key.encrypt(&mut rng, &too_big).is_err());
+        assert!(encryption_key.encrypt(&mut rng, &-too_big).is_err());
+
+        // Test a number bigger than N (both + and -)
+        let way_too_big = encryption_key.n() + 1;
+        assert!(encryption_key.encrypt(&mut rng, &way_too_big).is_err());
+        assert!(encryption_key.encrypt(&mut rng, &-way_too_big).is_err());
+    }
```

Comment:
Might be good to add tests right at the edge of the boundary to make sure there's no off-by-one happening.

Reply:
Added!